Change order of 'deleting' event

mellester opened the pull request:
It is my humble opinion that. The 'deleting' event should be fired before the item gets deleted from the database. Not after
Before this PR pivot::deleting would not have access to the Primary key  because its parent class usually model::delete already deleted so any listener will not have access to the full pivot table entry. because it would not exist in the database.

<!--
Please only send a pull request to branches which are currently supported: https://laravel.com/docs/releases#support-policy 

If you are unsure which branch your pull request should be sent to, please read: https://laravel.com/docs/contributions#which-branch

Pull requests without a descriptive title, thorough description, or tests will be closed.

In addition, please describe the benefit to end users; the reasons it does not break any existing features; how it makes building web applications easier, etc.
-->

mellester committed:
Change order of 'deleting' event

It is my humble opinion that. The 'deleting' event should be fired before the item gets deleted from the database. Not after
Before this PR pivot::deleting would not have access to the Primary key  because its parent class usually model::delete already deleted so any listener will not have access to the full pivot table entry. because it would not exist in the database.

mellester committed:
Whitespace fix

mellester committed:
ws fix

taylorotwell commented:
Did you actually have this problem in a real application? It calls `parent::delete` ... the parent's `delete` method will already firing the deleting event before deleting the record. With your change the `deleting` event would now be fired twice in this situation.

mellester commented:
Yeah @taylorotwell 
I had a custom trait which called `static::deleting`.
This event was passed a model with its primary key if and only if  the trait was added to a Class that Inherited from `Model`.
However if the trait was added to a `Pivot` class then the  `static::deleting` listener was not passed a model with the primary key.
And I could not retrieve it from the database.

Full disclosure I Was testing this on Laravel 7 not 8 .
And to fix this issue I simply directly called `model::deleting`
The bellow code works as intended 

```Php

    public static function bootRevisionableTrait()
    {
   

        static::created(function ($model) {
            dump($model->id, $model);

        });

        Model::deleting(function ($model) {
            dump('deleting', __LINE__, $model->id, $model);
         });

        Model::deleted(function ($model) {
            dd(__LINE__, $model->id, $model);
        });
```


However replacing `Model` with `static` doesn't  work as intended. 
$model->id was Null in the later case

In my use case I only needed to log the PrimaryKey of the item being deleted so the order of events doesn't matter to me.
but to others it might.
In my head I assumed the order of events would be like this, (not my use case)
Pivot::delleting -> First event to be fired,  A custom Listener could for example acquire Locks to all 3 tables, (pivot and foreign).
Model::delleting -> Second events to be fired, Checks to see if it has the lock to the (pivot table) or/and acquires it.
Model::deleted -> fired after the database entry was deleted, A custom listener could invalid the Current cache. and release locks
Pivot::deleted -> Last event to be fired ,  Release locks.

Currently the way I notice events to be fired in is as follows 
Model::delleting 
Model::deleted 
Pivot::delleting 
Pivot::deleted 

Whether this might be intended behavior and maybe allot of user are relying on this behavior.


However upon reading your comments and further review I might have misunderstood something.

I observed Pivot::delleting  not passing the Primary key and assumed the fault to be fixed by this PR but it doesn't 
It fires Multiple Events instead of addressing the underlying issue  being no pirmaryKey.
Thx for you time
